u2-routing-controllers/app.js

```diff
@@ -0,0 +1,31 @@
+const express = require('express')
+const { urlencoded, json } = require('body-parser')
+const { getUserByQuery, getUserByPath, getUserByForm, search } = require('./features/userController')
+const { logger } = require('./_services/logger')
+const { index } = require('./features/indexController')
+const { getProduct } = require('./features/productController')
+const { getRGB } = require('./features/rgbController')
+const { getTemp } = require('./features/tempController')
+const { getEcho, getEchoByForm } = require('./features/echoController')
+
+const app = express()
+
+// Middlewares
+app.use(urlencoded({ extended: false }))
+app.use(json()) // json() = function taht creates middleware
+app.use(logger) // logger = middleware
+
+// Routes
+app.get('/', index)
+app.get('/users', getUserByQuery)
+app.get('/users/search', search)
+app.get('/users/:username', getUserByPath)
+app.post('/users', getUserByForm)
+app.get('/products/search', getProduct)
+app.get('/rgb', getRGB)
+app.get('/:u1-to-:u2/:t1', getTemp)
+app.get('/echo', getEcho)
+app.post('/echo', getEchoByForm)
```

Comment:
ดี! สังเกต 3 กรณี:
- **req.query.xxx -- query string ที่ไม่ได้ประกาศใน `app.js`**  
  `/products/search?q=abc&page=4`  
  `/rgb?r=255&g=128&b=0`
- **req.params.xxx -- path parameters ประกาศใน `app.js` โดยใช้ `:`**  
  `/:u1-to-:u2/:t1`  
- **req.body.xxx -- ข้อมูลถูกส่งในเนื้อหาของคำขอ POST**  
  GET `/echo` ซึ่งให้บริการแบบฟอร์ม; แบบฟอร์มกำหนดชื่อของตัวแปร  
  POST `/echo` ซึ่งรับข้อมูลที่โพสต์